cranelift/codegen/src/opts/algebraic.isle

```diff
@@ -198,6 +196,30 @@
       (let ((mask Imm64 (imm64_shl ty (imm64 0xFFFF_FFFF_FFFF_FFFF) k)))
         (band ty x (iconst ty mask))))
 
+;; For unsigned shifts, `(x << k) >> k` is the same as masking out the top
+;; `k` bits. A similar rule is valid for vectors but this `iconst` mask only
+;; works for scalar integers.
+(rule (simplify (ushr (fits_in_64 (ty_int ty))
+                      (ishl ty x (iconst _ k))
+                      (iconst _ k)))
+      (band ty x (iconst ty (imm64_ushr ty (imm64 (ty_mask ty)) k))))
+
+
+;; Masking out any of the top bits of the result of `uextend` is a no-op. (This
+;; is like a cheap version of known-bits analysis.)
+(rule (simplify (band wide x @ (uextend _ (value_type narrow)) (iconst _ (u64_from_imm64 mask))))
+      ; This if-let is infallible, used for its let-binding effect.
+      (if-let narrow_mask (ty_mask narrow))
+      ; This if-let checks that `narrow_mask` has a subset of the bits that `mask` does.
+      (if-let narrow_mask (u64_and mask narrow_mask))
```

Comment:
(After talking with Jamey offline about this) it seems this is actually an equality constraint, counterintuitively! The `narrow_mask` is already defined and we start with the existing binding environment, not a fresh one, when building the AST for the LHS pattern; so we get the behavior one would for a second use of a variable in a pattern.

Here we should write an explicit equality (e.g. `(if-let $true (u64_eq ...))`) I think, and separately think about the above binding behavior...